common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -1042,4 +833,131 @@ public boolean contains(CharSequence cs) {
         }
         return indexOf(cs) >= 0;
     }
+
+    private static byte toLowerCase(byte b) {
+        if ('A' <= b && b <= 'Z') {
+            return (byte) (b + UPPER_CASE_TO_LOWER_CASE_ASCII_OFFSET);
+        }
+        return b;
+    }
+
+    private static char toLowerCase(char c) {
+        if ('A' <= c && c <= 'Z') {
+            return (char) (c + UPPER_CASE_TO_LOWER_CASE_ASCII_OFFSET);
+        }
+        return c;
+    }
+
+    private static byte toUpperCase(byte b) {
+        if ('a' <= b && b <= 'z') {
+            return (byte) (b - UPPER_CASE_TO_LOWER_CASE_ASCII_OFFSET);
+        }
+        return b;
+    }
+
+    /**
+     * Returns the case-insensitive hash code of the specified string. Note that this method uses the same hashing
+     * algorithm with {@link #hashCode()} so that you can put both {@link AsciiString}s and arbitrary
+     * {@link CharSequence}s into the same {@link TextHeaders}.
+     */
+    public static int caseInsensitiveHashCode(CharSequence value) {
+        if (value.getClass() == AsciiString.class) {
+            return ((AsciiString) value).hashCodeCaseInsensative();
+        }
+
+        int hash = 0;
+        for (int i = 0; i < value.length(); i++) {
+            hash = hash * HASH_CODE_PRIME + toLowerCase(value.charAt(i));
```

Comment:
@Scottmitch well it's always 31. but I can live with `HASH_CODE_PRIME`. Also can't we refactor the code to use the same algorithm, but in case of `AsciiString` we cache it? I know  bounds checks ... mucho expensive ... but I think simplicity of code is more important. Same goes for `hashCode()`.